### ISSUE TITLE
kube-proxy IPVS mode cannot be used due to missing ipset binary in hardened-kube-proxy image

### ISSUE THREAD

**Issue by myselfghost** [rancher/rke2#645]
kubectl -n kube-system logs kube-proxy-jzw8l:

```
W1224 17:45:15.895819       1 feature_gate.go:235] Setting GA feature gate DryRun=true. It will be removed in a future release.
W1224 17:45:15.895930       1 feature_gate.go:235] Setting GA feature gate RotateKubeletClientCertificate=true. It will beremoved in a future release.
W1224 17:45:15.895939       1 feature_gate.go:235] Setting GA feature gate EvenPodsSpread=true. It will be removed in a future release.
W1224 17:45:15.895945       1 feature_gate.go:235] Setting GA feature gate DefaultIngressClass=true. It will be removed ina future release.
W1224 17:45:15.896056       1 feature_gate.go:235] Setting GA feature gate DryRun=true. It will be removed in a future release.
W1224 17:45:15.896083       1 feature_gate.go:235] Setting GA feature gate RotateKubeletClientCertificate=true. It will beremoved in a future release.
W1224 17:45:15.896088       1 feature_gate.go:235] Setting GA feature gate EvenPodsSpread=true. It will be removed in a future release.
W1224 17:45:15.896095       1 feature_gate.go:235] Setting GA feature gate DefaultIngressClass=true. It will be removed ina future release.
E1224 17:45:15.904264       1 server_others.go:106] Can't use the IPVS proxier: error getting ipset version, error: executable file not found in $PATH
I1224 17:45:15.925678       1 node.go:136] Successfully retrieved node IP: 10.15.12.1
I1224 17:45:15.925711       1 server_others.go:142] kube-proxy node IP is an IPv4 address (10.15.12.1), assume IPv4 operation
I1224 17:45:15.954037       1 server_others.go:185] Using iptables Proxier.
I1224 17:45:15.954338       1 server.go:650] Version: v1.19.5
I1224 17:45:15.954679       1 conntrack.go:52] Setting nf_conntrack_max to 131072
I1224 17:45:15.954910       1 config.go:315] Starting service config controller
I1224 17:45:15.954927       1 shared_informer.go:240] Waiting for caches to sync for service config
I1224 17:45:15.955020       1 config.go:133] Starting endpoints config controller
I1224 17:45:15.955027       1 shared_informer.go:240] Waiting for caches to sync for endpoints config
I1224 17:45:16.055063       1 shared_informer.go:247] Caches are synced for service config
I1224 17:45:16.055185       1 shared_informer.go:247] Caches are synced for endpoints config

```

Can't use the IPVS proxier: error getting ipset version, error: executable file not found in $PATH

**Comment by rajivml** [comment]
am also running into this issue and more details about the same can be found here, because of this issue am **metalLB** LB isn't working as expect, the LB ip's that are getting generated, they are accessible only within the cluster but for them to be accessible out the cluster i.e. within the same subnet , kube-proxy has to be started in ipvs mode and I don't find any parameters through which kubeproxy can be started in ipvs mode.

https://rancher-users.slack.com/archives/C3ASABBD1/p1609243088298500

I tried changing ipvs mode in the kube-config configmap, but am getting these exceptions and kube-proxy is starting in "iptables proxy" mode itself 

```
root@rancher-federal-source:/home/sshuser/metallb# arp | grep 172.17.17
172.17.17.22                     (incomplete)                              eth0
[root@rancher-federal-airgap airgapped_install_rancher]# arping -I eth0 -c 3 172.17.17.22
ARPING 172.17.17.22 from 172.17.17.8 eth0
Sent 3 probes (3 broadcast(s))
Received 0 response(s)
[root@rancher-federal-airgap airgapped_install_rancher]# kubectl -n kube-system logs -f kube-proxy-48vbr
W1229 11:46:18.540123       1 proxier.go:635] Failed to load kernel module nf_conntrack_ipv4 with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
W1229 11:46:18.560926       1 proxier.go:635] Failed to load kernel module nf_conntrack_ipv4 with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
E1229 11:46:18.561036       1 server_others.go:569] can't determine whether to use ipvs proxy, error: error getting ipset version, error: executable file not found in $PATH
I1229 11:46:18.566751       1 node.go:136] Successfully retrieved node IP: 172.17.17.8
I1229 11:46:18.566779       1 server_others.go:186] Using iptables Proxier.
I1229 11:46:18.567061       1 server.go:583] Version: v1.18.13
```

**Comment by rajivml** [comment]
I went through this past issue which is more RKE specific, but I didn't understand how the steps mentioned over there will fit into RKE2 context .. It's better if there is an simpler out of box solution 

https://github.com/rancher/rancher/issues/13769

**Comment by bootc** [comment]
For people looking to make this fly, I managed to do it but had to use a 3rd party `kube-proxy` Docker image (more on that later):

```yaml
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-kube-proxy
  namespace: kube-system
spec:
  valuesContent: |-
    image:
      repository: gotok8s/kube-proxy
      tag: v1.20.4
    proxy:
      mode: ipvs
```

The custom image is required to work around this particular issue with `ipset` / `ipvsadm` not being available within the image. This is in fact just a mirror of the upstream Kubernetes `kube-proxy` image (unchanged from those published on `eu.gcr.io/k8s-artifacts-prod/kube-proxy`, I checked the manifest hash matches but don't trust me, check for yourself), but it's hosted on Docker Hub (because RKE2 hard-codes `docker.io` as a registry unless you set `system-default-registry` which affects far more than just `kube-proxy`).

So I think all it would take is for RKE2's `kube-proxy` image to include the necessary tools to allow this to work. In the mean time, I have a usable workaround.

**Comment by brandond** [comment]
FWIW, a fix for this needs to go into https://github.com/rancher/image-build-kube-proxy

**Comment by jniltinho** [comment]
> 
> 
> For people looking to make this fly, I managed to do it but had to use a 3rd party `kube-proxy` Docker image (more on that later):
> 
> ```yaml
> apiVersion: helm.cattle.io/v1
> kind: HelmChartConfig
> metadata:
>   name: rke2-kube-proxy
>   namespace: kube-system
> spec:
>   valuesContent: |-
>     image:
>       repository: gotok8s/kube-proxy
>       tag: v1.20.4
>     proxy:
>       mode: ipvs
> ```
> 
> The custom image is required to work around this particular issue with `ipset` / `ipvsadm` not being available within the image. This is in fact just a mirror of the upstream Kubernetes `kube-proxy` image (unchanged from those published on `eu.gcr.io/k8s-artifacts-prod/kube-proxy`, I checked the manifest hash matches but don't trust me, check for yourself), but it's hosted on Docker Hub (because RKE2 hard-codes `docker.io` as a registry unless you set `system-default-registry` which affects far more than just `kube-proxy`).
> 
> So I think all it would take is for RKE2's `kube-proxy` image to include the necessary tools to allow this to work. In the mean time, I have a usable workaround.

@bootc @brandond 

Can I put this config in that folder?
Is this how I can solve it now?
I am using the version of rke2 v1.18.17.

 ```
cd /var/lib/rancher/rke2/server/manifests/

cat > rke2-kube-proxy-metallb.yaml <<EOF
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-kube-proxy
  namespace: kube-system
spec:
  valuesContent: |-
    image:
      repository: gotok8s/kube-proxy
      tag: v1.20.4
    proxy:
      mode: ipvs
EOF

systemctl restart rke2-server
```

**Comment by brandond** [comment]
This will be fixed in the v1.20.6 release, if you can wait for that and save yourself some hacking.

**Comment by rancher-max** [comment]
#### Validated this is fixed in v1.20.5-alpha1+rke2r2

Note that version mentioned is just for testing sake and does not have bumped k8s versions yet and users should wait for the v1.20.6 release for this change.

I brought up my cluster with the following steps:
```
$ curl -sfL https://get.rke2.io | sudo INSTALL_RKE2_VERSION=v1.20.5-alpha1+rke2r2 INSTALL_RKE2_CHANNEL=testing sh -
$ sudo mkdir -p /var/lib/rancher/rke2/server/manifests
$ cat > /var/lib/rancher/rke2/server/manifests/rke2-kubeproxy-config.yaml <<EOF
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-kube-proxy
  namespace: kube-system
spec:
  valuesContent: |-
    proxy:
      mode: ipvs
EOF
$ sudo systemctl enable rke2-server.service && sudo systemctl start rke2-server.service
```
- Noticed no errors in the kube-proxy pod logs:
```
I0409 23:01:35.370557       1 node.go:172] Successfully retrieved node IP: 172.31.21.43
I0409 23:01:35.370643       1 server_others.go:142] kube-proxy node IP is an IPv4 address (172.31.21.43), assume IPv4 operation
I0409 23:01:35.423098       1 server_others.go:258] Using ipvs Proxier.
...
```
Previously using these same steps I noticed the error mentioned in this issue.

**Comment by harridu** [comment]
rke2 is version 1.21.7 now, but if I try to edit pod/kube-proxy-dpcl064.example.com to set
```
spec:
  containers:
  - command:
    - kube-proxy
    - --cluster-cidr=10.42.0.0/16
    - --conntrack-max-per-core=0
    - --conntrack-tcp-timeout-close-wait=0s
    - --conntrack-tcp-timeout-established=0s
    - --healthz-bind-address=127.0.0.1
    - --hostname-override=dpcl064.ac.aixigo.de
    - --kubeconfig=/var/lib/rancher/rke2/agent/kubeproxy.kubeconfig
    - --proxy-mode=ipvs
    - --ipvs-strict-arp
    env:
    - name: FILE_HASH
```
then it is refused:
```
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
# pods "kube-proxy-dpcl064.ac.aixigo.de" was not valid:
# * spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`, `spec.initContainers[*].image`, `spec.activeDeadlineSeconds` or `spec.tolerations` (only additions to existing tolerations)
#   core.PodSpec{
#   	Volumes:        {{Name: "file0", VolumeSource: {HostPath: &{Path: "/var/lib/rancher/rke2/agent/client-kube-proxy.crt", Type: &"File"}}}, {Name: "file1", VolumeSource: {HostPath: &{Path: "/var/lib/rancher/rke2/agent/client-kube-proxy.key", Type: &"File"}}}, {Name: "file2", VolumeSource: {HostPath: &{Path: "/var/lib/rancher/rke2/agent/kubeproxy.kubeconfig", Type: &"File"}}}, {Name: "file3", VolumeSource: {HostPath: &{Path: "/var/lib/rancher/rke2/agent/server-ca.crt", Type: &"File"}}}},
#   	InitContainers: nil,
#   	Containers: []core.Container{
#   		{
#   			Name:  "kube-proxy",
#   			Image: "index.docker.io/rancher/hardened-kubernetes:v1.21.7-rke2r2-build"...,
#   			Command: []string{
#   				... // 6 identical elements
#   				"--hostname-override=dpcl064.ac.aixigo.de",
#   				"--kubeconfig=/var/lib/rancher/rke2/agent/kubeproxy.kubeconfig",
# - 				"--proxy-mode=ipvs",
# + 				"--proxy-mode=iptables",
# - 				"--ipvs-strict-arp",
#   			},
#   			Args:       nil,
#   			WorkingDir: "",
#   			... // 17 identical fields
#   		},
#   	},
#   	EphemeralContainers: nil,
#   	RestartPolicy:       "Always",
#   	... // 25 identical fields
#   }
```
